### PR TITLE
py domain: Add py:property directive to describe a property (refs: #7068)

### PR DESCRIPTION
**Pull request by tk0miya**
### Feature or Bugfix
- Feature
- Refactoring

### Purpose
- refs: #7068

### CODE REVIEW

File: sphinx/domains/python.py [hunk 1]
```diff
@@ -674,6 +674,8 @@ def get_signature_prefix(self, sig: str) -> str:
         if 'classmethod' in self.options:
             prefix.append('classmethod')
         if 'property' in self.options:
+            logger.warning(__('Use the directive `py:property` instead of '
+                              'the directive `py:method` for properties.'))
```
Comment (tk0miya) [inline sphinx/domains/python.py]: Hmm... it seems it is better to separate deprecation to another PR to avoid breaking tests.

File: doc/usage/restructuredtext/domains.rst [hunk 2]
```diff
@@ -253,6 +253,22 @@ The following directives are provided for module and class contents:
...
+   Describes an object property.
+
+   .. versionadded:: 3.0
+
+   .. rubric:: options
+
+   .. rst:directive:option:: abstractmethod
+      :type: no value
+
+      Indicate the property is an abstract method.
```
Comment (flying-sheep) [inline doc/usage/restructuredtext/domains.rst]:
```suggestion
      Indicate the property is abstract.
```

Properties aren’t methods, they just happen to use `@abstractmethod`

File: doc/usage/restructuredtext/domains.rst [hunk 3]
```diff
@@ -253,6 +253,22 @@ The following directives are provided for module and class contents:
...
+
+   .. rst:directive:option:: type: type of the property
+      :type: text
```
Comment (flying-sheep) [inline doc/usage/restructuredtext/domains.rst:333]: Maybe we should also add the type the setter accepts? E.g. I have a property that can accept any `typing.PathLike` but always coerces it to `pathlib.Path`, so you can do:

```py
settings.path = './foo'
assert settings.path == Path('./foo')

File: doc/usage/restructuredtext/domains.rst [hunk 4]
```diff
@@ -506,6 +526,10 @@ a matching identifier is found:
 
    Reference a data attribute of an object.
 
+.. rst:role:: py:prop
+
+   Reference a property of an object.
```
Comment (flying-sheep) [inline doc/usage/restructuredtext/domains.rst]:
```suggestion
   Reference a property or attribute of an object.
```

Since both can do both, this should also be the new description of `:attr:`

File: sphinx/domains/python.py [hunk 5]
```diff
@@ -1159,8 +1202,14 @@ def resolve_xref(self, env: BuildEnvironment, fromdocname: str, builder: Builder
...
         if not matches and type == 'attr':
-            # fallback to meth (for property)
+            # fallback to prop (for property)
+            matches = self.find_obj(env, modname, clsname, target, 'prop', searchmode)
+        if not matches and type == 'attr':
+            # fallback to meth (for property; Sphinx-2.4.x)
             matches = self.find_obj(env, modname, clsname, target, 'meth', searchmode)
+        if not matches and type == 'meth':
+            # fallback to prop (for property)
+            matches = self.find_obj(env, modname, clsname, target, 'prop', searchmode)
```
Comment (flying-sheep) [inline sphinx/domains/python.py]: You also need to add the following, right?

Since people can transparently change attributes to properties, too?

```py
if not matches and type == 'prop':
    matches = self.find_obj(env, modname, clsname, target, 'attr', searchmode)
```

And maybe more informative comments would be:

1. for attr/prop:
   
   ```py
   # properties and attributes are transparently interchangeable,
   # so their roles need to be able to refer to the respective other
   ```

2. for `meth`:

   ```py
   # fallback to meth (in Sphinx 2.4.x, properties were stored as py:method)
   ```

File: tests/test_domain_py.py [hunk 6]
```diff
@@ -178,6 +180,10 @@ def test_resolve_xref_for_properties(app, status, warning):
             ' title="module_a.submodule.ModTopLevel.prop">'
             '<code class="xref py py-meth docutils literal notranslate"><span class="pre">'
             'prop</span> <span class="pre">method</span></code></a>' in content)
+    assert ('Link to <a class="reference internal" href="#module-a-submodule-modtoplevel-prop"'
+            ' title="module_a.submodule.ModTopLevel.prop">'
+            '<code class="xref py py-prop docutils literal notranslate"><span class="pre">'
```
Comment (flying-sheep) [inline tests/test_domain_py.py]:
```suggestion
            '<code class="xref py py-attr py-prop docutils literal notranslate"><span class="pre">'
```

Backwards compatibility for themes without sacrificing the ability to style both differently.

**Review by flying-sheep — COMMENTED** [review]
A few changes to address the concerns raised in #7068

File: doc/usage/restructuredtext/domains.rst (hunk 3 above)
Comment (eric-wieser) [reply doc/usage/restructuredtext/domains.rst:333]: Not sure how you'd want to write or display that. Perhaps the folks over at mypy can recommend a spelling?

Comment (flying-sheep) [reply doc/usage/restructuredtext/domains.rst:333]: In code it’s obvious to express this: A return type annotation for the getter and a parameter type annotation on the setter.

In rST it could be `:get-type:` or `:type:`, and if a setter exists, also `:set-type:`.

I don’t know a concise way to display it (like “**Return Type**” for functions), except for spelling out “Any $Type can be assigned to it”

Comment (eric-wieser) [reply doc/usage/restructuredtext/domains.rst:333]: Perhaps if `typing.Descriptor` could be added, as was suggested in the comment thread 
 at https://github.com/python/mypy/issues/244#issuecomment-620113658, then this could just be written `:type: Descriptor[GetType, SetType]`.

Comment (tk0miya) [reply doc/usage/restructuredtext/domains.rst:333]: I think this should be an another topic. The goal of this PR is replacing property-description to `py:property` directive.

File: sphinx/domains/python.py (hunk 5 above)
Comment (tk0miya) [reply sphinx/domains/python.py]: Never discussed that `py:prop` role can refer to attributes. And I don't think so.

Comment (flying-sheep) [reply sphinx/domains/python.py]: It should, as people can always also transparently change properties to attributes. It’s going to happen rarely, but it will.

Comment (eric-wieser) [reply sphinx/domains/python.py]: I haven't followed the full discussion here, but I think I side with @flying-sheep here - in my opinion either:

* Only `:attr:` should exist
* `:attr:` / `:prop:` should be completely interchangeable

I'd argue that while these's some value to describing `.. attribute` and `.. property` differently 
 to draw attention to dynamic behavior, there's no value in referencing them differently, as to a consumer there's no difference.

Comment (flying-sheep) [reply sphinx/domains/python.py]: That’s been my argument all along. @tk0miya thinks the consistency of having a one role per directive is important, while I tended towards going more in the “Only `:attr:` should exist” direction, because of the Python zen: “there should be one – and preferably only one – obvious way to do it”. But @tk0miya and me had a loooong discussion that we don’t want to repeat so I think the best compromise is “`:attr:` / `:prop:` should be completely interchangeable”

File: sphinx/domains/python.py [hunk 7]
```diff
@@ -887,6 +887,46 @@ def get_index_text(self, modname: str, name_cls: Tuple[str, str]) -> str:
...
+            clsname, attrname = name.rsplit('.', 1)
+            if modname and self.env.config.add_module_names:
+                clsname = '.'.join([modname, clsname])
+        except ValueError:
+            if modname:
+                return _('%s (in module %s)') % (name, modname)
+            else:
+                return name
+
+        return _('%s (%s property)') % (attrname, clsname)
```
Comment (eric-wieser) [inline sphinx/domains/python.py:865]: Better to keep as much as possible out of the `try`:
```suggestion
        except ValueError:
            if modname:
                return _('%s (in module %s)') % (name, modname)
            else:
                return name

        if modname and self.env.config.add_module_names:
            clsname = '.'.join([modname, clsname])

        return _('%s (%s property)') % (attrname, clsname)
```

File: sphinx/domains/python.py [hunk 8]
```diff
@@ -1016,6 +1056,23 @@ def process_link(self, env: BuildEnvironment, refnode: Element,
         return title, target
 
 
+class PyPropXRefRole(PyXRefRole):
+    """A special xref role class for :py:prop: role.
+
+    It adds a "py-attr" style to the role for backward compatibility.
+
+    Note: RemovedInSphinx50Warning
```
Comment (eric-wieser) [inline sphinx/domains/python.py]: @flying-sheep: This isn't about backwards-compatibility of _sphinx_, right? It's about backwards compatibility of libraries changing between `attr` and `prop` - so I don't think it makes sense to ever deprecate it.

File: sphinx/domains/python.py (hunk 5 above)
Comment (tk0miya) [reply sphinx/domains/python.py]: @eric-wieser Who is the "producer" (provider?) and "consumer"?

I think no reason to describe them differently if they are transparent. Is it worthy to describe them differently? If so, they're different things, I think.

Comment (eric-wieser) [reply sphinx/domains/python.py]: To be clear, the "provider" is the docs that use `.. py:property`, such as a framework that provides the property. A "consumer" is a docstring in a docstream package that wants to reference the property.

> I think no reason to describe them differently if they are transparent. Is it worthy to describe them differently? If so, they're different things, I think.

Let's use `:abstract:` as an example - it's useful for readers to know that a method is abstract when reading its docstring, but when linking `` :abstract-meth:`some_meth` `` wouldn't be useful.

So by that reasoning I'd argue that `:property:` and `:abstract:` should be treated the same way:
```
.. method:: my_abstract_meth
    :abstract:

    This is the doc for :meth:`my_abstract_meth`, which I do not mention
    the abstract-ness of when I link.

.. attribute:: my_prop
    :property:

    This is the doc for :attr:`my_prop`, which I do not mention
    the property-ness of when I link.
```
I realize I'm at risk of reviving @flying-sheep's discussion here though.

Comment (tk0miya) [reply sphinx/domains/python.py]: Do you mean to use `py:attribute` directive to describe property (as an option)? If so, I can understand your idea. But I choose to add a new type `py:property` because it is unnatural to describe "abstract" or "async" properties via `py:attribute` directive. Additionally, client library sometimes needs to know the "member" of object is an attribute or a property (ex. assigning value, inheritance, etc).

Should we discuss "what is a property"?

File: sphinx/domains/python.py (hunk 8 above)
Comment (flying-sheep) [reply sphinx/domains/python.py]: Yes, it’s about backwards compat of themes and custom css, and it doesn’t make sense to remove it. Most themes will probably forever style both kinds of links the same way.

In fact, I can’t think of a reason not to style the links the same way.

File: sphinx/domains/python.py (hunk 7 above)
Comment (eric-wieser) [reply sphinx/domains/python.py:865]: I think this comment still applies?

**Comment by tk0miya** [comment]
Finally, I determined to drop the `py:prop` role because I felt tired to discuss more.

File: sphinx/domains/python.py (hunk 5 above)
Comment (eric-wieser) [reply sphinx/domains/python.py]: > Do you mean to use py:attribute directive to describe property (as an option)

Yes, I guess I was suggesting that

> because it is unnatural to describe "abstract" or "async" properties

The former is `@abstractproperty`, and the latter has a PyPI package; https://pypi.org/project/async-property/.

So I think actually that's an argument _for_ `:property:` just being metadata used in `..attribute`

File: sphinx/domains/python.py [hunk 9]
```diff
@@ -1046,6 +1086,7 @@ class PythonDomain(Domain):
         'classmethod':  ObjType(_('class method'),  'meth', 'obj'),
         'staticmethod': ObjType(_('static method'), 'meth', 'obj'),
         'attribute':    ObjType(_('attribute'),     'attr', 'obj'),
+        'property':     ObjType(_('property'),      'prop', 'attr', 'obj'),
```
Comment (eric-wieser) [inline sphinx/domains/python.py]: I thought you said you had removed the `prop` role - but isn't that what this line is defining? Or do I misunderstand?

Comment (tk0miya) [reply sphinx/domains/python.py]: What do you mean? This is a definition of object-type, not a role.

Comment (eric-wieser) [reply sphinx/domains/python.py]: The documentation of `ObjType` says the arguments are:

> *roles*: all the roles that can refer to an object of this type

Comment (tk0miya) [reply sphinx/domains/python.py]: Ah, okay. understand.

File: sphinx/domains/python.py [hunk 10]
```diff
@@ -1195,8 +1237,11 @@ def resolve_xref(self, env: BuildEnvironment, fromdocname: str, builder: Builder
                                 type, searchmode)
 
         if not matches and type == 'attr':
-            # fallback to meth (for property)
+            # fallback to meth (for property; Sphinx-2.4.x)
             matches = self.find_obj(env, modname, clsname, target, 'meth', searchmode)
+        if not matches and type == 'meth':
+            # fallback to attr (for property)
+            matches = self.find_obj(env, modname, clsname, target, 'attr', searchmode)
```
Comment (tk0miya) [inline sphinx/domains/python.py]: As a fallback for `meth` role, python domain searches the property using `attr` role. So it might hit an incorrect one.

Comment (eric-wieser) [reply sphinx/domains/python.py]: So as I understand it; the first `if` is for:

* Users who still use `..method` to define their properties
* Users loading an intersphinx file generated by an older sphinx that write `meth` to `objects.inv`.

While the latter `if` is for:

* Users who still use `:meth:` to refer to properties

Does it make sense to add a check that the object being referred to is actually a property?

Comment (tk0miya) [reply sphinx/domains/python.py]: I think a "prop" role is a polite way to do that. But it seems all of you don't like to introduce it. Please don't ask me.

Comment (eric-wieser) [reply sphinx/domains/python.py]: Ah, I finally understand what you're getting at. There's no way for sphinx to know if `` :meth:`maybe_prop` `` is a property  or not, as it can't see the old-style `:property:` marker; that information doesn't make it into `objects.inv`.

What I think I'm suggesting is that a warning should be emitted. I'll make another comment with a better set of lines selected...

Comment (eric-wieser) [reply sphinx/domains/python.py]: Actually, this is what I'd suggest:
```suggestion
            # this ensures that `:meth:` in old rst files can refer to `..property` in new rst files.
            matches = self.find_obj(env, modname, clsname, target, 'attr', searchmode)
            # don't allow `:meth:` ever to refer to regular attributes.
            matches = [match for match in matches if match[1] == 'property']
```

File: sphinx/domains/python.py [hunk 11]
```diff
@@ -1195,8 +1237,11 @@ def resolve_xref(self, env: BuildEnvironment, fromdocname: str, builder: Builder
                                 type, searchmode)
 
         if not matches and type == 'attr':
-            # fallback to meth (for property)
+            # fallback to meth (for property; Sphinx-2.4.x)
```
Comment (eric-wieser) [inline sphinx/domains/python.py:1250]:
```suggestion
            # fallback to meth (for property; Sphinx-2.4.x)
            # this ensures that `:attr:` continues to refer to the deprecated `.. method::` `:property:` in
            # old rst files.
```

File: sphinx/domains/python.py (hunk 10 above)
Comment (tk0miya) [reply sphinx/domains/python.py]: Why do you reinvent the prop role? Is it better to implement it for internal use?

Comment (eric-wieser) [reply sphinx/domains/python.py]: Adding a prop role wouldn't help, because this `if` block is dealing with old rST files which are incorrectly using a `:meth:` role. If you add a new `:prop:` role, these old files will still incorrectly use a `:meth:` role.

Conversely, new rST files have nothing to gain from a `:prop:` role, as the existing `:attr:` role is all they need.

The messiness of the code I suggest is simply due to trying to be backwards compatible. In sphinx 5, you could remove both of these if statements if you wanted.

Comment (tk0miya) [reply sphinx/domains/python.py]: I'd like to say use a prop role as a fallback of the `:meth:` role in the old reST files. It behaves the same as what your suggested code doing.
```
            matches = self.find_obj(env, modname, clsname, target, 'attr', searchmode)
            matches = [match for match in matches if match[1] == 'property']
```
and
```
            matches = self.find_obj(env, modname, clsname, target, 'prop', searchmode)
```
will be equivalent.

Comment (eric-wieser) [reply sphinx/domains/python.py]: Am I right in thinking that with that change, users will be able to type `` :prop:`...` `` in their rST files though, which is what we were trying to avoid? Or does "role" mean two different things here, one internal to sphinx and one exposed through rST

If the latter, how about `_prop_backcompat` to make it clear that this is for internal use?

Comment (tk0miya) [reply sphinx/domains/python.py]: Yes, I meant the latter one. I feel `_prop_backcompat` is too much. Adding a comment is enough helping us.

Comment (eric-wieser) [reply sphinx/domains/python.py]: How about just `_prop` then with a comment?

Comment (tk0miya) [reply sphinx/domains/python.py]: Okay, go with this.

**Comment by tk0miya** [comment]
Merged. Thank you for your comments!
